Firestore/core/src/local/local_documents_view.cc

```diff
@@ -141,79 +119,212 @@ model::DocumentMap LocalDocumentsView::GetDocumentsMatchingCollectionGroupQuery(
+  // include documents for all overlays in the initial document set.
+  for (const auto& entry : overlays) {
+    if (remote_documents.find(entry.first) == remote_documents.end()) {
+      remote_documents = remote_documents.insert(
+          entry.first, MutableDocument::InvalidDocument(entry.first));
+    }
+  }
 
-  for (const MutationBatch& batch : matching_batches) {
-    for (const Mutation& mutation : batch.mutations()) {
-      // Only process documents belonging to the collection.
-      if (!query.path().IsImmediateParentOf(mutation.key().path())) {
-        continue;
-      }
+  // Apply the overlays and match against the query.
+  DocumentMap results;
+  for (const auto& entry : remote_documents) {
+    const auto& key = entry.first;
+    MutableDocument doc = entry.second;
 
-      const DocumentKey& key = mutation.key();
-      // base_doc may be unset for the documents that weren't yet written to
-      // the backend.
-      absl::optional<MutableDocument> document = remote_documents.get(key);
-      if (!document) {
-        // Create invalid document to apply mutations on top of
-        document = MutableDocument::InvalidDocument(key);
-      }
+    auto overlay_it = overlays.find(key);
+    if (overlay_it != overlays.end()) {
+      (*overlay_it)
+          .second.mutation()
+          .ApplyToLocalView(doc, FieldMask(), Timestamp::Now());
+    }
+    // Finally, insert the documents that still match the query
+    if (query.Matches(doc)) {
+      results = results.insert(key, std::move(doc));
+    }
+  }
+
+  return results;
+}
+
+Document LocalDocumentsView::GetDocument(const DocumentKey& key) {
+  absl::optional<Overlay> overlay = document_overlay_cache_->GetOverlay(key);
+  MutableDocument document = GetBaseDocument(key, overlay);
+  if (overlay.has_value()) {
+    overlay.value().mutation().ApplyToLocalView(document, FieldMask(),
+                                                Timestamp::Now());
+  }
+  return document;
+}
 
-      // TODO(Overlay): Use proper previous mask.
-      mutation.ApplyToLocalView(*document, absl::nullopt,
-                                batch.local_write_time());
-      remote_documents = remote_documents.insert(key, *document);
+DocumentMap LocalDocumentsView::GetDocuments(const DocumentKeySet& keys) {
+  MutableDocumentMap docs = remote_document_cache_->GetAll(keys);
+  return GetLocalViewOfDocuments(docs, DocumentKeySet{});
+}
+
+DocumentMap LocalDocumentsView::GetLocalViewOfDocuments(
+    const MutableDocumentMap& base_docs,
+    const DocumentKeySet& existence_state_changed) {
+  OverlayByDocumentKeyMap overlays;
+  PopulateOverlays(overlays, DocumentKeySet::FromKeysOf(base_docs));
+  auto overlayed_documents =
+      ComputeViews(base_docs, std::move(overlays), existence_state_changed);
+
+  DocumentMap result;
+  for (auto& entry : overlayed_documents) {
+    result =
+        result.insert(entry.first, std::move(entry.second.consume_document()));
```

Comment:
Remove std::move here... it is not necessary because `consume_document()` returns a temporary.

Just for completeness, I'd like to reiterate my preference for `std::move(entry.second).document()` here :)

Reply:
I really do not have a strong preference, but I see the similarities with optional now, so let's make it rvalue ref.

Reply:
I pushed up a commit to tweak to add the proper lvalue/rvalue ref-qualified overloads of `document()`.